### PR TITLE
Introduce Dynamo Service Mode as new CLI flag

### PR DESCRIPTION
**Pull request by QilongTang**
Please Note:
1. Before submitting the PR, please review [How to Contribute to Dynamo](https://github.com/DynamoDS/Dynamo/blob/master/CONTRIBUTING.md)
2. Dynamo Team will meet 1x a month to review PRs found on Github (Issues will be handled separately)
3. PRs will be reviewed from oldest to newest
4. If a reviewed PR requires changes by the owner, the owner of the PR has 30 days to respond. If the PR has seen no activity by the next session, it will be either closed by the team or depending on its utility will be taken over by someone on the team
5. PRs should use either Dynamo's default PR template or [one of these other template options](https://github.com/DynamoDS/Dynamo/wiki/Choosing-a-Pull-Request-Template) in order to be considered for review.
6. PRs that do not have one of the Dynamo PR templates completely filled out with all declarations satisfied will not be reviewed by the Dynamo team.
7. PRs made to the `DynamoRevit` repo will need to be cherry-picked into all the DynamoRevit Release branches that Dynamo supports. Contributors will be responsible for cherry-picking their reviewed commits to the other branches after a `LGTM` label is added to the PR.

### Purpose

This is a PR introducing service mode `-s` as CLI flag to minimize code startup time. I have finished going through most of the initialization logic and decided to skip certain ones with graph opening tested.

Here is the current list of steps that can be skipped:
DynamoModel:
- [x] Reading ExeConfiguration about `DisableAnalytics` flag for Alias Team 
- [x] Analytics Client initialization
- [x] Dynamo settings migrator
- [x] Adding default trust locations
- [x] Python template loading
- [x] Search Model
- [x] LinterManager initialization
- [x] AuthenticationManager initialization
- [x] UpdateManager.CheckForProductUpdate Call (I assume this call is expensive that connecting with AWS bucket to get latest Dynamo Release info)
- [x] Package Manager extension loading

DynamoViewModel:

- [x] Incanvas Search ViewModel initialization
- [x] Node AutoComplete Search ViewModel initialization
- [x] Node display in the library

image

Performance gain:
Test env: Macbook M1 - MacOS - Parallel - Win 11 setup
The memory diff of launching DynamoWPF CLI in keep-alive mode (what player is using) 
Before: ~116mb

After: ~104mb roughly 10% gain.
image

The profiling time (DotTrace - Trace Mode) diff of launching DynamoWPF CLI in keep-alive mode (what player is using) 

Before: ~10300ms

After: ~7900ms with roughly ~24% gain in the current state of PR. I suggest Team Morpheus wrap up the current work and introduce the service mode officially.
image

### Declarations

Check these if you believe they are true

- [x] The codebase is in a better state after this PR
- [x] Is documented according to the [standards](https://github.com/DynamoDS/Dynamo/wiki/Coding-Standards)
- [x] The level of testing this PR includes is appropriate
- [ ] User facing strings, if any, are extracted into `*.resx` files
- [x] All tests pass using the self-service CI.
- [x] Snapshot of UI changes, if any.
- [x] Changes to the API follow [Semantic Versioning](https://github.com/DynamoDS/Dynamo/wiki/Dynamo-Versions) and are documented in the [API Changes](https://github.com/DynamoDS/Dynamo/wiki/API-Changes) document.
- [ ] This PR modifies some build requirements and the readme is updated

### Release Notes

Introduce Dynamo Service Mode for CLI usage and minimal cold start time.

### Reviewers

@DynamoDS/dynamo 

### FYIs

@Amoursol

### CODE REVIEW

File: src/DynamoApplications/StartupUtils.cs [hunk 1]
```diff
@@ -220,7 +221,9 @@ public static CommandLineArguments Parse(string[] args)
                 .Add("si=|SI=|sessionId", "Identify Dynamo host analytics session id", si => sessionId = si)
                 .Add("pi=|PI=|parentId", "Identify Dynamo host analytics parent id", pi => parentId = pi)
                 .Add("da|DA|disableAnalytics", "Disables analytics in Dynamo for the process liftime", da => disableAnalytics = da != null)
-                .Add("cr=|CR=|cerLocation", "Specify the crash error report tool location on disk ", cr => cerLocation = cr);
+                .Add("cr=|CR=|cerLocation", "Specify the crash error report tool location on disk ", cr => cerLocation = cr)
+                .Add("s|S|service mode", "Service mode, by pass certain Dynamo launch steps for maximum startup performance", s => serviceMode = s != null);
```
Comment (mjkkirschner) [inline src/DynamoApplications/StartupUtils.cs:225]: spelling (bypass one word)

File: src/DynamoCore/Models/DynamoModel.cs [hunk 2]
```diff
@@ -699,9 +704,10 @@ protected DynamoModel(IStartConfiguration config)
                 // Do nothing for now
             }
 
+
             // If user skipped analytics from assembly config, do not try to launch the analytics client
             // or the feature flags client.
-            if (!areAnalyticsDisabledFromConfig && !Dynamo.Logging.Analytics.DisableAnalytics)
+            if (!areAnalyticsDisabledFromConfig && !Dynamo.Logging.Analytics.DisableAnalytics && !isServiceMode)
```
Comment (QilongTang) [inline src/DynamoCore/Models/DynamoModel.cs:710]: @reddyashish @zeusongit This is purely a kick-start, let's brainstorm what can be skipped in service mode. Feel free to touch this branch with your ideas.

Comment (mjkkirschner) [reply src/DynamoCore/Models/DynamoModel.cs:710]: good idea to validate these choices with profiling I guess.

Comment (QilongTang) [reply src/DynamoCore/Models/DynamoModel.cs]: @saintentropy @mjkkirschner @BogdanZavu Any of you have profiled our WPFCLI? Any tips?

File: test/DynamoCoreTests/AuthenticationManagerTests.cs [hunk 3]
```diff
@@ -74,15 +74,6 @@ public void ToggleLoginState()
...
         }
 
-        [Test]
-        public void CanToggleLoginState()
-        {
-            var authProviderMock = new Mock<IAuthProvider>();
-            var authManager = new AuthenticationManager(authProviderMock.Object);
-
-            Assert.IsTrue(authManager.CanToggleLoginState());
-        }
```
Comment (QilongTang) [inline test/DynamoCoreTests/AuthenticationManagerTests.cs:84]: @reddyashish Does this test fail only in service mode or? Would you share the reason we remove it?

Comment (reddyashish) [reply test/DynamoCoreTests/AuthenticationManagerTests.cs:84]: @QilongTang This change was made by Filipe here https://github.com/DynamoDS/Dynamo/pull/13657, to remove legacy code. We need to sync this branch with master.

Comment (QilongTang) [reply test/DynamoCoreTests/AuthenticationManagerTests.cs:84]: @reddyashish Thanks!

File: src/DynamoCoreWpf/Views/Core/DynamoView.xaml.cs [hunk 4]
```diff
@@ -911,15 +911,18 @@ private void DynamoViewModelRequestViewOperation(ViewOperationEventArgs e)
 
         private void DynamoLoadedViewExtensionHandler(ViewLoadedParams loadedParams, IEnumerable<IViewExtension> extensions)
         {
-            foreach (var ext in extensions)
+            if (!dynamoViewModel.Model.IsServiceMode)
```
Comment (QilongTang) [inline src/DynamoCoreWpf/Views/Core/DynamoView.xaml.cs:914]: @reddyashish This would not be triggered because Dynamo View is not initialized in WPFCLI mode or CLI mode. I suspect there are other places we launch view extensions. Will try to locate, I may revert this change

Comment (reddyashish) [reply src/DynamoCoreWpf/Views/Core/DynamoView.xaml.cs:914]: yeah, my understanding was that these view extensions are loaded after the view is initialized. I will look too if there is any other place where it is launched.

File: src/DynamoCore/Models/DynamoModel.cs [hunk 5]
```diff
@@ -942,14 +942,13 @@ protected DynamoModel(IStartConfiguration config)
             DynamoModel.OnRequestUpdateLoadBarStatus(new SplashScreenLoadEventArgs(Resources.SplashScreenLoadNodeLibrary, 50));
             InitializeNodeLibrary();
 
-            if (!IsServiceMode && extensions.Any())
+            if (extensions.Any())
```
Comment (reddyashish) [inline src/DynamoCore/Models/DynamoModel.cs:945]: this is to address dynamo player case?

Comment (QilongTang) [reply src/DynamoCore/Models/DynamoModel.cs:945]: yeah I am uncertain if this is needed or not because @saintentropy mentioned that Player will be loaded as an extension, but it seems we can still skip loading package manager?

File: src/DynamoCoreWpf/Views/Core/DynamoView.xaml.cs (hunk 4 above)
Comment (QilongTang) [reply src/DynamoCoreWpf/Views/Core/DynamoView.xaml.cs:914]: @reddyashish Gotcha, no wonder the DynamoViewModel initialization did not impact launch time that much

File: src/DynamoCore/Models/DynamoModel.cs (hunk 5 above)
Comment (QilongTang) [reply src/DynamoCore/Models/DynamoModel.cs]: @reddyashish Adding this back to skip loading package manager as we planned, when needed, we can add logic for loading player specifically

File: src/DynamoApplications/StartupUtils.cs [hunk 6]
```diff
@@ -279,6 +283,11 @@ private static void ShowHelp(OptionSet opSet)
             public bool DisableAnalytics { get; set; }
             public HostAnalyticsInfo AnalyticsInfo { get; set; }
             public string CERLocation { get; set; }
+
+            /// <summary>
+            /// Boolean indication of launching Dynamo in service mode, this mode it optimized for minimal launch time
```
Comment (mjkkirschner) [inline src/DynamoApplications/StartupUtils.cs:288]: spelling (is) optimized

File: src/DynamoApplications/StartupUtils.cs [hunk 7]
```diff
@@ -220,7 +221,9 @@ public static CommandLineArguments Parse(string[] args)
                 .Add("si=|SI=|sessionId", "Identify Dynamo host analytics session id", si => sessionId = si)
                 .Add("pi=|PI=|parentId", "Identify Dynamo host analytics parent id", pi => parentId = pi)
                 .Add("da|DA|disableAnalytics", "Disables analytics in Dynamo for the process liftime", da => disableAnalytics = da != null)
-                .Add("cr=|CR=|cerLocation", "Specify the crash error report tool location on disk ", cr => cerLocation = cr);
+                .Add("cr=|CR=|cerLocation", "Specify the crash error report tool location on disk ", cr => cerLocation = cr)
+                .Add("s|S|service mode", "Service mode, bypass certain Dynamo launch steps for maximum startup performance", s => serviceMode = s != null);
```
Comment (mjkkirschner) [inline src/DynamoApplications/StartupUtils.cs:225]: spelling (bypasses)

File: src/DynamoApplications/StartupUtils.cs [hunk 8]
```diff
@@ -455,19 +484,23 @@ private static bool PreloadASM(string asmPath, out string geometryFactoryPath, o
...
                 ProcessMode = CLImode ? TaskProcessMode.Synchronous : TaskProcessMode.Asynchronous,
                 HostAnalyticsInfo = info,
-                CLIMode = CLImode
+                CLIMode = CLImode,
+                AuthProvider = CLImode ? null : new Core.IDSDKManager(),
+                UpdateManager = CLImode ? null : OSHelper.IsWindows() ? InitializeUpdateManager() : null,
+                StartInTestMode = CLImode,
+                PathResolver = CLImode ? new CLIPathResolver(preloaderLocation, userDataFolder, commonDataFolder) as IPathResolver : new SandboxPathResolver(preloaderLocation) as IPathResolver,
+                // TODO: Dynamo 3.0
+                // Update Default start config to include IsServiceMode so DynamoModel can be initialize with that flag. For now, reuse the legacy Context property
```
Comment (mjkkirschner) [inline src/DynamoApplications/StartupUtils.cs:501]: you can safely modify DefaultStartConfig to include IsServiceMode you just can't change the interface - so you can add a new interface.

File: src/DynamoCore/Models/DynamoModel.cs [hunk 9]
```diff
@@ -208,6 +208,11 @@ public string Version
         /// </summary>
         public HostAnalyticsInfo HostAnalyticsInfo { get; set; }
 
+        /// <summary>
+        /// Boolean indication of launching Dynamo in service mode, this mode it optimized for minimal launch time, mostly leveraged by CLI or WPF CLI.
```
Comment (mjkkirschner) [inline src/DynamoCore/Models/DynamoModel.cs:212]: is optimized

File: src/DynamoCore/Models/DynamoModel.cs [hunk 10]
```diff
@@ -208,6 +208,11 @@ public string Version
...
+        /// </summary>
+        public bool IsServiceMode { get; set; }
```
Comment (mjkkirschner) [inline src/DynamoCore/Models/DynamoModel.cs:214]: should the setter really be public?

File: src/DynamoApplications/StartupUtils.cs [hunk 11]
```diff
@@ -339,6 +349,25 @@ public static DynamoModel MakeCLIModel(string asmPath, string userDataFolder, st
...
+        /// <summary>
+        /// Use this overload to construct a DynamoModel in CLI context when the location of ASM to use is known, host analytics info is known and you want to set data paths.
+        /// </summary>
+        /// <param name="asmPath">Path to directory containing geometry library binaries</param>
+        /// <param name="userDataFolder">Path to be used by PathResolver for UserDataFolder</param>
+        /// <param name="commonDataFolder">Path to be used by PathResolver for CommonDataFolder</param>
+        /// <param name="info">Host analytics info specifying Dynamo launching host related information.</param>
+        /// <param name="isServiceMode">Boolean indication of launching Dynamo in service mode, this mode is optimized for minimal launch time.</param>
+        /// <returns></returns>
+        public static DynamoModel MakeCLIModel(string asmPath, string userDataFolder, string commonDataFolder, HostAnalyticsInfo info = new HostAnalyticsInfo(), bool isServiceMode = false)
```
Comment (BogdanZavu) [inline src/DynamoApplications/StartupUtils.cs:361]: I have also added a couple of flags last year :). I guess we need to think of something soon to better consolidate all the MakeModel constructs. Perhaps a startup config file where you can enable/disable features/extensions etc.

File: src/DynamoCore/Models/DynamoModel.cs [hunk 12]
```diff
@@ -682,26 +693,29 @@ protected DynamoModel(IStartConfiguration config)
...
-            catch (Exception)
-            {
-                // Do nothing for now
             }
 
             // If user skipped analytics from assembly config, do not try to launch the analytics client
-            // or the feature flags client.
-            if (!areAnalyticsDisabledFromConfig && !Dynamo.Logging.Analytics.DisableAnalytics)
+            // or the feature flags client for web traffic reason.
+            if (!areAnalyticsDisabledFromConfig && !Dynamo.Logging.Analytics.DisableAnalytics && !IsServiceMode)
```
Comment (BogdanZavu) [inline src/DynamoCore/Models/DynamoModel.cs:718]: We can put the IsServiceMode check at the beginning of these conditions.

File: src/DynamoCore/Models/DynamoModel.cs [hunk 13]
```diff
@@ -848,19 +863,29 @@ protected DynamoModel(IStartConfiguration config)
...
 
             NodeFactory = new NodeFactory();
             NodeFactory.MessageLogged += LogMessage;
 
             //Initialize the ExtensionManager with the CommonDataDirectory so that extensions found here are checked first for dll's with signed certificates
             extensionManager = new ExtensionManager(new[] { PathManager.CommonDataDirectory });
             extensionManager.MessageLogged += LogMessage;
-            var extensions = config.Extensions ?? LoadExtensions();
+            var extensions = config.Extensions ?? new List<IExtension>();
+            if (!extensions.Any() && !IsServiceMode)
```
Comment (BogdanZavu) [inline src/DynamoCore/Models/DynamoModel.cs:880]: Is there a list somewhere with all our default extensions and what are they doing ?

File: src/DynamoCoreWpf/ViewModels/Core/DynamoViewModelDelegateCommands.cs [hunk 14]
```diff
@@ -65,12 +65,17 @@ private void InitializeDelegateCommands()
...
             GettingStartedGuideCommand = new DelegateCommand(StartGettingStartedGuide, CanStartGettingStartedGuide);
             ShowPackageManagerSearchCommand = new DelegateCommand(ShowPackageManagerSearch, CanShowPackageManagerSearch);
-            PublishNewPackageCommand = new DelegateCommand(PackageManagerClientViewModel.PublishNewPackage, PackageManagerClientViewModel.CanPublishNewPackage);
             ShowInstalledPackagesCommand = new DelegateCommand(o => { }, o => true);
             ManagePackagePathsCommand = new DelegateCommand(o => { }, o => true);
-            PublishCurrentWorkspaceCommand = new DelegateCommand(PackageManagerClientViewModel.PublishCurrentWorkspace, PackageManagerClientViewModel.CanPublishCurrentWorkspace);
-            PublishSelectedNodesCommand = new DelegateCommand(PackageManagerClientViewModel.PublishSelectedNodes, PackageManagerClientViewModel.CanPublishSelectedNodes);
-            PublishCustomNodeCommand = new DelegateCommand<Function>(PackageManagerClientViewModel.PublishCustomNode, PackageManagerClientViewModel.CanPublishCustomNode);
+
+            if (PackageManagerClientViewModel != null)
```
Comment (BogdanZavu) [inline src/DynamoCoreWpf/ViewModels/Core/DynamoViewModelDelegateCommands.cs:71]: Should we continue to check for IsServiceMode as well ? If we're not in service mode and PackageManagerClientViewModel is null I guess we are in an erroneous situation ?

File: src/WorkspaceDependencyViewExtension/WorkspaceDependencyViewExtension.cs [hunk 15]
```diff
@@ -84,11 +84,13 @@ public override void Loaded(ViewLoadedParams viewLoadedParams)
             // as we may have installed a missing package.
 
             DependencyView.CustomNodeManager = (CustomNodeManager)viewLoadedParams.StartupParams.CustomNodeManager;
-
-            pmExtension.PackageLoader.PackgeLoaded += (package) =>
+            if (pmExtension != null)
```
Comment (BogdanZavu) [inline src/WorkspaceDependencyViewExtension/WorkspaceDependencyViewExtension.cs:87]: If pmExtenions does not exist does it makes sense to load this one or we can simply return here at the beginning ?

Comment (QilongTang) [reply src/WorkspaceDependencyViewExtension/WorkspaceDependencyViewExtension.cs:87]: There are other types of workspace dependencies (e.g. local dependency, custom node workspace dependency) that are not depending on pmExtension, so it still make sense to load I believe.

File: src/DynamoCore/Models/DynamoModel.cs (hunk 13 above)
Comment (QilongTang) [reply src/DynamoCore/Models/DynamoModel.cs:880]: In code, we dont maintain a list. We only maintain the list using xml manifest files in `extensions` and `viewextensions` folder

File: src/DynamoCore/Models/DynamoModel.cs (hunk 12 above)
Comment (QilongTang) [reply src/DynamoCore/Models/DynamoModel.cs:718]: Will do

File: src/DynamoApplications/StartupUtils.cs (hunk 11 above)
Comment (QilongTang) [reply src/DynamoApplications/StartupUtils.cs:361]: agreed, currently the code (constructors) used for CLI are a bit messy

File: src/DynamoCoreWpf/ViewModels/Core/DynamoViewModelDelegateCommands.cs (hunk 14 above)
Comment (QilongTang) [reply src/DynamoCoreWpf/ViewModels/Core/DynamoViewModelDelegateCommands.cs:71]: Can do

**Comment by QilongTang** [comment]
@mjkkirschner @BogdanZavu @reddyashish @zeusongit @pinzart @aparajit-pratap @sm6srw All tests are passing so if you would like to make this `service mode` official for CLI, we can do that as well now.
